core/shared/src/main/scala/zio/Config.scala

```diff
@@ -0,0 +1,289 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zio
+
+import scala.util.control.{NonFatal, NoStackTrace}
+
+/**
+ * A [[zio.Config]] describes the structure of some configuration data.
+ */
+sealed trait Config[+A] { self =>
+
+  /**
+   * Returns a new config that is the composition of this config and the
+   * specified config.
+   */
+  def ++[B](that: Config[B])(implicit zippable: Zippable[A, B]): Config[zippable.Out] =
+    Config.Zipped[A, B, zippable.Out](self, that, zippable)
+
+  /**
+   * Returns a config whose structure is preferentially described by this
+   * config, but which falls back to the specified config if there is an issue
+   * reading from this config.
+   */
+  def ||[A1 >: A](that: Config[A1]): Config[A1] = Config.Fallback(self, that)
+
+  /**
+   * Adds a description to this configuration, which is intended for humans.
+   */
+  def ??(label: String): Config[A] = Config.Described(self, label)
+
+  /**
+   * Returns a new config whose structure is the same as this one, but which
+   * produces a different Scala value, constructed using the specified function.
+   */
+  def map[B](f: A => B): Config[B] = self.mapOrFail(a => Right(f(a)))
+
+  /**
+   * Returns a new config whose structure is the samea as this one, but which
+   * may produce a different Scala value, constructed using the specified
+   * fallible function.
+   */
+  def mapOrFail[B](f: A => Either[Config.Error, B]): Config[B] = Config.MapOrFail(self, f)
+
+  /**
+   * Returns a new config whose structure is the same as this one, but which may
+   * produce a different Scala value, constructed using the specified function,
+   * which may throw exceptions that will be translated into validation errors.
+   */
+  def mapOrThrow[B](f: A => B): Config[B] =
+    self.mapOrFail { a =>
+      try Right(f(a))
+      catch {
+        case NonFatal(e) => Left(Config.Error.InvalidData(Chunk.empty, e.getMessage))
+      }
+    }
+
+  /**
+   * Returns a new config that has this configuration nested as a property of
+   * the specified name.
+   */
+  def nested(name: String): Config[A] = Config.Nested(name, self)
+
+  /**
+   * Returns an optional version of this config, which will be `None` if the
+   * data is missing from configuration, and `Some` otherwise.
+   */
+  def optional: Config[Option[A]] = self.map(Some(_)) || Config.succeed(None)
+
+  /**
+   * A named version of `||`.
+   */
+  def orElse[A1 >: A](that: Config[A1]): Config[A1] = self || that
```

Comment:
Isn't it better to make the alphanum functions the canonical representations? The symbolic operators can be just (sometimes handy) aliases.

Reply:
Yes, but that's not very standard in ZIO, see Schedule. We don't use operators very often, but when we do use them, we encourage their use.